### PR TITLE
`<tuple>` enabled on gcc and clang 

### PR DESCRIPTION
**Pull request by wmaxey**
Builds are clean on nvrtc, win32, and major gcc and clang versions.

MSVC is disabled.

### CODE REVIEW

**Comment by brycelelbach** [comment]
@wmaxey let's ship `<tuple>` for non-MSVC compilers this week in 1.3.0 for the next CUDA toolkit.

**Comment by wmaxey** [comment]
> @wmaxey let's ship `<tuple>` for non-MSVC compilers this week in 1.3.0 for the next CUDA toolkit.

@brycelelbach 
Sounds good.

I've noticed that the reference construction issue is present in gcc-7.5 similar to gcc-8. 

Tests are exempt on MSVC so it should be good to go there. Should we try to apply a workaround for these compilers?

Re: What we uncovered in August: there might be a few approaches that would satisfy NVCC's default reference construction complaint.

**Comment by brycelelbach** [comment]
> I've noticed that the reference construction issue is present in gcc-7.5 similar to gcc-8.

Do we have a workaround for that?

**Comment by brycelelbach** [comment]
> Re: What we uncovered in August: there might be a few approaches that would satisfy NVCC's default reference construction complaint.

Can we implement any of them this week?

**Comment by wmaxey** [comment]
I've committed the changes that WAR the issue for gcc-8 and added a test to demonstrate if the fix is no longer required.

**Comment by wmaxey** [comment]
@brycelelbach @griwes This is ready to be merged in. Just need a review to satisfy GH if I'm not mistaken.

File: libcxx/include/tuple [hunk 1]
```diff
@@ -223,12 +224,13 @@ public:
...
-              class = _EnableIf<
+              _EnableIf<
                   _And<
                       _IsNotSame<__uncvref_t<_Tp>, __tuple_leaf>,
                       is_constructible<_Hp, _Tp>
-                    >::value
-                >
+                    >::value,
+                  bool
+                > = false
```
Comment (griwes) [inline libcxx/include/tuple:233]: I'm really curious why this change was needed.

File: libcxx/include/tuple [hunk 2]
```diff
@@ -939,10 +963,11 @@ public:
...
 
     template <class _Tuple,
-              class = typename enable_if
+              typename enable_if
                       <
-                         __tuple_assignable<_Tuple, tuple>::value
-                      >::type
+                         __tuple_assignable<_Tuple, tuple>::value,
+                         void*
+                      >::type = nullptr
```
Comment (griwes) [inline libcxx/include/tuple:970]: A clever trick here is to not add the second argument, but add a `*` after `::type`. But this is fine too ;)

File: libcxx/include/tuple [hunk 3]
```diff
@@ -1132,7 +1156,11 @@ struct __ignore_t
 };
 
 namespace {
+#ifdef __CUDA_ARCH__
+  _LIBCUDACXX_INLINE_VAR __constant__ __ignore_t<unsigned char> ignore = __ignore_t<unsigned char>();
+#else
   _LIBCUDACXX_INLINE_VAR constexpr __ignore_t<unsigned char> ignore = __ignore_t<unsigned char>();
+#endif
```
Comment (griwes) [inline libcxx/include/tuple:1163]: A similar - though not identical - pattern also appeared in the complex changes; I wonder if there's a nice abstraction we could have for when we need to do selection of essentially a single token like this.

**Review by griwes — APPROVED** [review]
Minor comments, but don't gate shipping this on them.

File: libcxx/include/tuple (hunk 1 above)
Comment (wmaxey) [reply libcxx/include/tuple:233]: I've seen less SFINAE issues with MSVC when using value assignment rather than types. This may have been done out of an abundance of caution.

File: libcxx/include/tuple (hunk 2 above)
Comment (wmaxey) [reply libcxx/include/tuple:970]: I'll update this to an actual `_EnableIf` from the internal meta functions.

File: libcxx/include/tuple (hunk 3 above)
Comment (wmaxey) [reply libcxx/include/tuple:1163]: I'll turn this into an issue. :)
